core/reorder/mc64.cpp

```diff
@@ -0,0 +1,621 @@
+                marked_cols, matched_idxs, queue, q_j, tolerance);
+        }
+        root = unmatched[um];
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+void compute_scaling(const matrix::Csr<ValueType, IndexType>* host_mtx,
+                     const array<remove_complex<ValueType>>& weights_array,
+                     const array<remove_complex<ValueType>>& dual_u_array,
+                     const array<remove_complex<ValueType>>& row_maxima_array,
+                     const array<IndexType>& permutation,
+                     const array<IndexType>& matched_idxs_array,
+                     mc64_strategy strategy, ValueType* row_scaling,
+                     ValueType* col_scaling)
+{
+    constexpr auto inf =
+        std::numeric_limits<remove_complex<ValueType>>::infinity();
+    const auto num_rows = host_mtx->get_size()[0];
+    const auto weights = weights_array.get_const_data();
+    const auto dual_u = dual_u_array.get_const_data();
+    const auto row_maxima = row_maxima_array.get_const_data();
+    const auto p = permutation.get_const_data();
+    const auto idxs = matched_idxs_array.get_const_data();
+
+    if (strategy == mc64_strategy::max_diagonal_product) {
+        for (size_type i = 0; i < num_rows; i++) {
+            const remove_complex<ValueType> u_val = std::exp2(dual_u[i]);
+            const remove_complex<ValueType> v_val =
+                std::exp2(weights[idxs[i]] - dual_u[p[i]] - row_maxima[i]);
+            col_scaling[i] = ValueType{u_val};
+            row_scaling[i] = ValueType{v_val};
+        }
+    } else {
+        for (size_type i = 0; i < num_rows; i++) {
+            col_scaling[i] = 1.;
+            row_scaling[i] = 1.;
+        }
+    }
+}
+
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_MC64_INITIALIZE_WEIGHTS);
+GKO_INSTANTIATE_FOR_EACH_NON_COMPLEX_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_MC64_INITIAL_MATCHING);
+GKO_INSTANTIATE_FOR_EACH_NON_COMPLEX_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_MC64_SHORTEST_AUGMENTING_PATH);
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(GKO_DECLARE_MC64_COMPUTE_SCALING);
+
+
+}  // namespace mc64
+
+
+namespace {
+
+
+GKO_REGISTER_HOST_OPERATION(initialize_weights, mc64::initialize_weights);
+GKO_REGISTER_HOST_OPERATION(initial_matching, mc64::initial_matching);
+GKO_REGISTER_HOST_OPERATION(augment_matching, mc64::augment_matching);
+GKO_REGISTER_HOST_OPERATION(compute_scaling, mc64::compute_scaling);
+GKO_REGISTER_OPERATION(fill_seq_array, components::fill_seq_array);
+
+
+}  // namespace
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<Composition<ValueType>> Mc64<ValueType, IndexType>::generate(
+    std::shared_ptr<const LinOp> system_matrix) const
+{
+    auto product = std::unique_ptr<Composition<ValueType>>(
+        static_cast<Composition<ValueType>*>(
+            this->LinOpFactory::generate(std::move(system_matrix)).release()));
```

Comment:
why does it need to go through LinOpFactory::generate?

Reply:
Otherwise we would recursively call this `generate` function, leading to infinite recursion. Instead we need to call the base class function which in turn calls `generate_impl`, and cast the result to the correct type.